rust/wal3/src/manifest.rs

```diff
@@ -0,0 +1,557 @@
+            >= snapshot_options.snapshot_rollover_threshold;
+        let can_snapshot_fragments =
+            self.fragments.len() >= snapshot_options.fragment_rollover_threshold;
+        if can_snapshot_snapshots || can_snapshot_fragments {
+            // NOTE(rescrv):  We _either_ compact a snapshot of snapshots or a snapshot of log
+            // fragments.  We don't do both as interior snapshot nodes only refer to objects of the
+            // same type.  Manifests are the only objects to refer to both fragments and snapshots.
+            let mut snapshots = vec![];
+            let mut fragments = vec![];
+            let mut setsum = Setsum::default();
+            let depth = if can_snapshot_snapshots {
+                for snapshot in self.snapshots.iter() {
+                    if snapshot.depth < 2
+                        && snapshots.len() < snapshot_options.snapshot_rollover_threshold
+                    {
+                        setsum += snapshot.setsum;
+                        snapshots.push(snapshot.clone());
+                    }
+                }
+                2
+            } else if can_snapshot_fragments {
+                for fragment in self.fragments.iter() {
+                    // NOTE(rescrv):  When taking a snapshot, it's important that we keep around
+                    // one fragment so that the max seq no is always calculable.
+                    //
+                    // Otherwise, a low-traffic log could be compacted into a state where all of
+                    // its fragments have been compacted and therefore the implicit fragment seq no
+                    // for each fragment is zero.  This wedges the manifest manager.
+                    //
+                    // The fix is to keep around the last fragment.
+                    if fragments.len() < snapshot_options.fragment_rollover_threshold
+                        && self
+                            .fragments
+                            .iter()
+                            .map(|f| f.seq_no)
+                            .max()
+                            .unwrap_or(FragmentSeqNo(0))
+                            != fragment.seq_no
+                    {
+                        setsum += fragment.setsum;
+                        fragments.push(fragment.clone());
+                    }
+                }
+                1
+            } else {
+                unreachable!();
+            };
+            let path = snapshot_path(setsum);
+            Some(Snapshot {
+                path,
+                depth,
+                setsum,
+                writer,
+                snapshots,
+                fragments,
+            })
+        } else {
+            None
+        }
+    }
+
+    /// Given a snapshot, apply it to the manifest.  This modifies the manifest to refer to the
+    /// snapshot and removes from the snapshot all data that is now part of the snapshot.
```

Comment:
> removes from the snapshot all data that is now part of the snapshot.

Should this be 

"removes from the **manifest** all data that is now part of the snapshot."

?